keps/sig-storage/1432-volume-health-monitor/README.md

```diff
@@ -674,8 +679,200 @@ _This section must be completed when targeting alpha to a release._
+  Longer term, we may want to require automated upgrade/rollback tests, but we
+  are missing a bunch of machinery and tooling and can't do that now.
+  Manual testing will be done.
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
+fields of API types, flags, etc.?**
+  Even if applying deprecation policies, they may still surprise some users.
+  No.
+
+### Monitoring Requirements
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How can an operator determine if the feature is in use by workloads?**
+  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+  checking if there are objects with field X set) may be a last resort. Avoid
+  logs or events for this purpose.
+  An operator can check the metric `csi_sidecar_operations_seconds`,
+  Container Storage Interface operation duration with gRPC error code status
+  total. It is reported from CSI external-health-monitor-controller and
+  external-health-monitor-agent sidecars. For the health monitor controller
+  sidecar, `csi_sidecar_operations_seconds` will be measuring `ListVolumes` or
+  `GetVolume` RPC. For the health monitor agent sidecar,
+  `csi_sidecar_operations_seconds` will be measuring `NodeGetVolumeStats` RPC.
+  The `csi_sidecar_operations_seconds` metric should be sliced by process after
+  they are aggregated to show metrics for different sidecars.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [ ] Metrics
+    - Metric name: csi_sidecar_operations_seconds
+    - [Optional] Aggregation method:
+    - Components exposing the metric: csi-external-health-monitor-controller and csi-external-health-monitor-agent sidecars
+  - [ ] Other (treat as last resort)
+    - Details:
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+  At a high level, this usually will be in the form of "high percentile of SLI
+  per day <= X". It's impossible to provide comprehensive guidance, but at the very
+  high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99,9% of /health requests per day finish with 200 code
+
+  The metrics `csi_sidecar_operations_seconds` includes a gRPC status code. If the
+  status code is `OK`, the call is successful; otherwise, it is not successful. We
+  can look at the ratio of successful vs non-successful statue codes to figure out
+  the success/failure ratio.
+
+* **Are there any missing metrics that would be useful to have to improve observability
+of this feature?**
+  Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+  implementation difficulties, etc.).
```

Comment:
Are we producing operation counts in addition to durations? (I'm guessing yes if `csi_sidecar_operations_seconds` is a histogram but we might want to note the existence such metrics in here explicitly.)

Reply:
Yes the histogram includes duration, counts and return code.